src/future_result.rs

```diff
@@ -0,0 +1,129 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Poll;
+
+use crate::TantivyError;
+
+/// `FutureResult` is a handle that makes it possible to wait for the completion
+/// of an ongoing task.
+///
+/// Contrary to some `Future`, it does not need to be polled for the task to
+/// progress. Dropping the `FutureResult` does not cancel the task being executed
+/// either.
+///
+/// - In a sync context, you can call `FutureResult::wait()`. The function
+/// does not rely on `block_on`.
+/// - In an async context, you can call simply use `FutureResult` as a future.
+pub struct FutureResult<T> {
+    inner: Inner<T>,
+}
+
+enum Inner<T> {
+    FailedBeforeStart(Option<TantivyError>),
+    InProgress {
+        receiver: oneshot::Receiver<crate::Result<T>>,
+        error_msg_if_failure: &'static str,
+    },
+}
+
+impl<T> From<TantivyError> for FutureResult<T> {
+    fn from(err: TantivyError) -> Self {
+        FutureResult {
+            inner: Inner::FailedBeforeStart(Some(err)),
+        }
+    }
+}
+
+impl<T> FutureResult<T> {
+    pub(crate) fn create(
+        error_msg_if_failure: &'static str,
+    ) -> (Self, oneshot::Sender<crate::Result<T>>) {
+        let (sender, receiver) = oneshot::channel();
+        let inner: Inner<T> = Inner::InProgress {
+            receiver,
+            error_msg_if_failure,
+        };
+        (FutureResult { inner }, sender)
+    }
+
+    /// Blocks until the scheduled result is available.
+    ///
+    /// In an async context, you should simply use `ScheduledResult` as a future.
+    pub fn wait(self) -> crate::Result<T> {
+        match self.inner {
+            Inner::FailedBeforeStart(err) => Err(err.unwrap()),
+            Inner::InProgress {
+                receiver,
+                error_msg_if_failure,
+            } => receiver.recv().unwrap_or_else(|_| {
+                Err(crate::TantivyError::SystemError(
+                    error_msg_if_failure.to_string(),
+                ))
```

Comment:
can't we return the original error?

Reply:
We do if on the other side, we sent the error to the FutureResult.
This error is only used if the sender side was dropped.

Even if we were very careful it could happen on a panic for instance.